jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/publishers/JacocoReportPublisher.java

```diff
@@ -127,6 +127,8 @@ public void process(@Nonnull StepContext context, @Nonnull Element mavenSpyLogsE
         Launcher launcher = context.get(Launcher.class);
 
         List<Element> jacocoPrepareAgentEvents = XmlUtils.getExecutionEventsByPlugin(mavenSpyLogsElt, "org.jacoco", "jacoco-maven-plugin", "prepare-agent", "MojoSucceeded", "MojoFailed");
+        List<Element> jacocoPrepareAgentEvents2 = XmlUtils.getExecutionEventsByPlugin(mavenSpyLogsElt, "org.jacoco", "jacoco-maven-plugin", "prepare-agent-integration", "MojoSucceeded", "MojoFailed");
```

Comment:
Hi @marioja 

Thanks for this PR

I agree that we don't support the "prepare-agent-integration" and we should probably do it if someone is using only IT tests and not Unit tests

Maybe you could rename `jacocoPrepareAgentEvents2` to `jacocoPrepareAgentIntegrationEvents` for a better understanding.

I think it my be great to add also a test case

You can maybe create `maven_jar_with_jacoco_integration_project` from https://github.com/jenkinsci/pipeline-maven-plugin/tree/master/jenkins-plugin/src/test/resources/org/jenkinsci/plugins/pipeline/maven/test/test_maven_projects/maven_jar_with_jacoco_project and adapt it for integration tests using failsafe

Then you have to copy https://github.com/jenkinsci/pipeline-maven-plugin/blob/master/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/AbstractIntegrationTest.java#L73

And write a new test similar to https://github.com/jenkinsci/pipeline-maven-plugin/blob/master/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepOnMasterTest.java#L187-L227

WDYT ?

Reply:
I agree with the variable name change.  I think that adding the prepare-agent-integration to the current maven_jar_with_jacoco_project and adding a new test in WithMavenStepOnMasterTest.java is probably a better approach.  I am working on this as we speak (or type 👍).

Reply:
Not sure though this should be labeled as enhancement.  Not including integration test code coverage I think is more a bug.  I don't know how significant it is to label it enhancement or not.  Also, contrary to your initial comment I don't think it is nice to have integration tests in case there are no junit but rather I think it is necessary to include integration tests whenever they exist. WDYT?

Reply:
I always hesitate for such kind of issue if this is a missing feature or a bug. I agree in any case that we should support it.
My idea to add another project was to be sure that all combinations (UT, IT, UT+IT) are working well. it might be painful to write 3 test cases and I was thinking that UT + IT in different projects should provide a better coverage than a unique UT+IT which could hide some others cases.

@bguerin any thoughts ?

Reply:
Well, we could have 3 tests: UT, IT and UT+IT but I think that a single UT+IT test should give us the same information (i.e. UT JaCoCo file is produced and IT JaCoCo file is produced) we would get if we did UT and IT separately.  I do not think it is important to "verify" that IT JaCoCo coverage file is not present when we don't run IT tests.  Same for UT.  If they were present this would have to be because we invoked the test and our test setup is wrong and in that case even that would not be too consequential because the functionality we really want tested is that if you have UT test, a UT file is produced and published and if a IT test is run, a IT file is produced and published.  My 2 cents (even though it is more like the length of a $20).

Reply:
let's do one test with UT+IT @marioja 
I think you are right, it should protect us enough

Reply:
I spent the better part of today trying to do the JUnit test case but I found the following problems (in order of importance):

1. The only jacoco JUnit test case (WithMavenStepOnMasterTest#maven_build_jar_with_jacoco_succeeds) is actually ignored(probably for a good reason)
2. When I tried to run it, I realized that it expects to run on Linux.  I am on Windows.  I tried to run it in a docker but I got all kinds of setup problems and sorted them up (which took too long)
3. The last error I got broke my back:
`java.lang.IncompatibleClassChangeError: class org.jacoco.core.internal.flow.ClassProbesVisitor has interface org.objectweb.asm.ClassVisitor as super class
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
        at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
        at hudson.plugins.jacoco.ExecutionFileLoader.analyzeStructure(ExecutionFileLoader.java:109)
        at hudson.plugins.jacoco.ExecutionFileLoader.loadBundleCoverage(ExecutionFileLoader.java:140)
        at hudson.plugins.jacoco.JacocoReportDir.parse(JacocoReportDir.java:110)
        at hudson.plugins.jacoco.JacocoBuildAction.loadRatios(JacocoBuildAction.java:331)
        at hudson.plugins.jacoco.JacocoBuildAction.load(JacocoBuildAction.java:321)
        at hudson.plugins.jacoco.JacocoPublisher.perform(JacocoPublisher.java:675)
        at org.jenkinsci.plugins.pipeline.maven.publishers.JacocoReportPublisher.process(JacocoReportPublisher.java:219)
        at org.jenkinsci.plugins.pipeline.maven.MavenSpyLogProcessor.processMavenSpyLogs(MavenSpyLogProcessor.java:153)
        at org.jenkinsci.plugins.pipeline.maven.WithMavenStepExecution2$WithMavenStepExecutionCallBack.finished(WithMavenStepExecution2.java:1097)
        at org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution$TailCall.lambda$onSuccess$0(GeneralNonBlockingStepExecution.java:140)
        at org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution.lambda$run$0(GeneralNonBlockingStepExecution.java:77)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE`

I am going to submit the code change and hopefully it can be incorporated quickly in a new version especially since the current jacoco support is not covered by a JUnit case at the present.

Reply:
Sorry that you lost so much time on this @marioja 
I will look at it. I think our tests should pass on windows (at least they do on the CI) but I am not a windows expert and I couldn't be surprised by some specific environment issues.
I will try to have a look at it and build the testcase. I cannot promise how quickly it could be done but I will do my best.

Reply:
When I run on windows it complains about command nohup not being available.  I think the CI build is running on linux (docker maybe).

Can we have a new version of the pipeline-maven before you attempt the JUnit test case?  We are kinda waiting for this on our Jenkins CI infrastructure and I would like to use a "release" plugin.

I can even help you doing the JUnit test case, I would like to learn how it all works as I might want to use this in the future.

Reply:
Actually, I may have found other bugs.  Hold on.

Reply:
Trying to find an additional bug (when destFile element not specified in jacoco plugin config) I discovered that in order to make the JUnit test work on windows you need to "fix" the use of sh in about 5 dynamically created pipelines during the test runs.  Also, I found out that if you want reproduceable results in Eclipse you must run `mvn install`  to run the test.  `mvn test` does not cut it.  Also, I found out that I get the following AssertError when running under Windows:
```
java.lang.AssertionError
	at hudson.FilePath$ExplicitlySpecifiedDirScanner.scan(FilePath.java:3146)
	at hudson.FilePath$CopyRecursiveLocal.invoke(FilePath.java:2433)
	at hudson.FilePath$CopyRecursiveLocal.invoke(FilePath.java:2417)
	at hudson.FilePath.act(FilePath.java:1075)
	at hudson.FilePath.act(FilePath.java:1058)
	at hudson.FilePath.copyRecursiveTo(FilePath.java:2365)
	at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:70)
	at org.jenkinsci.plugins.pipeline.maven.publishers.GeneratedArtifactsPublisher.process(GeneratedArtifactsPublisher.java:118)
	at org.jenkinsci.plugins.pipeline.maven.MavenSpyLogProcessor.processMavenSpyLogs(MavenSpyLogProcessor.java:153)
	at org.jenkinsci.plugins.pipeline.maven.WithMavenStepExecution2$WithMavenStepExecutionCallBack.finished(WithMavenStepExecution2.java:1097)
	at org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution$TailCall.lambda$onSuccess$0(GeneralNonBlockingStepExecution.java:140)
	at org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution.lambda$run$0(GeneralNonBlockingStepExecution.java:77)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```
Why on earth would the hudson.FilePath$ExplicitlySpecifiedDirScanner.scan method crash on path with back slash in them?  I can only surmise that this is the case because there is no code in hudson that uses that ExplicitlySpecifiedDirScanner class except for in these JUnit test cases.  This is where I stopped trying on Windows.  I am not about to create another pull for hudson to "fix" this.  I went out to create a Ubuntu Hyper-V VM (another endeavor) and I was able to run the JUnit test reliably using `mvn install` but it was so late that I did not have time to test any of my jacoco changes nor create meaningful JUnit test cases for them.

Reply:
Here is a git patch that "documents" the changes I did on Windows which were not sufficient to make it work:
```
diff --git a/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepGlobalConfigurationTest.java b/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepGlobalConfigurationTest.java
index a240f9c..651a511 100644
--- a/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepGlobalConfigurationTest.java
+++ b/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepGlobalConfigurationTest.java
@@ -86,7 +86,7 @@
             String pipelineScript = "node('master') {\n" +
                     "    git($/" + gitRepoRule.toString() + "/$)\n" +
                     "    withMaven() {\n" +
-                    "        sh 'mvn package verify'\n" +
+                    "        bat 'mvn package verify'\n" +
                     "    }\n" +
                     "}";
 
diff --git a/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepOnMasterTest.java b/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepOnMasterTest.java
index 94c5667..826b8dc 100644
--- a/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepOnMasterTest.java
+++ b/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepOnMasterTest.java
@@ -98,7 +98,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0') {\n" +
-                "        sh 'mvn package'\n" +
+                "        bat 'mvn package'\n" +
                 "    }\n" +
                 "}";
 
@@ -125,7 +125,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven(maven: 'install-does-not-exist') {\n" +
-                "        sh 'mvn package'\n" +
+                "        bat 'mvn package'\n" +
                 "    }\n" +
                 "}";
 
@@ -141,7 +141,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven() {\n" +
-                "        sh 'mvn package verify'\n" +
+                "        bat 'mvn package verify'\n" +
                 "    }\n" +
                 "}";
 
@@ -200,7 +200,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven() {\n" +
-                "        sh 'mvn package verify'\n" +
+                "        bat 'mvn package verify'\n" +
                 "    }\n" +
                 "}";
 
@@ -234,7 +234,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven() {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}";
 
@@ -325,7 +325,7 @@
             String pipelineScript = "node('master') {\n" +
                     "    git($/" + gitRepoRule.toString() + "/$)\n" +
                     "    withMaven(options:[" + symbol + "(disabled:" + disabled + ")]) {\n" +
-                    "        sh 'mvn package verify'\n" +
+                    "        bat 'mvn package verify'\n" +
                     "    }\n" +
                     "}";
 
@@ -363,7 +363,7 @@
                 "       pattern:'src/main/java', excludePattern:'a/path'," +
                 "       ignoreCase:true, asRegexp:false, " +
                 "       lowPriorityTaskIdentifiers:'minor', normalPriorityTaskIdentifiers:'todo', highPriorityTaskIdentifiers:'fixme')]) {\n" +
-                "           sh 'mvn package verify'\n" +
+                "           bat 'mvn package verify'\n" +
                 "    }\n" +
                 "}";
 
@@ -382,7 +382,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven() {\n" +
-                "        sh 'mvn package'\n" +
+                "        bat 'mvn package'\n" +
                 "    }\n" +
                 "}";
 
@@ -424,7 +424,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven() {\n" +
-                "        sh 'mvn package'\n" +
+                "        bat 'mvn package'\n" +
                 "    }\n" +
                 "}";
 
@@ -464,7 +464,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven() {\n" +
-                "        sh 'mvn package'\n" +
+                "        bat 'mvn package'\n" +
                 "    }\n" +
                 "}";
 
@@ -506,7 +506,7 @@
         String pipelineScript = "node('master') {\n" +
                 "    git($/" + gitRepoRule.toString() + "/$)\n" +
                 "    withMaven() {\n" +
-                "        sh 'mvn test'\n" +
+                "        bat 'mvn test'\n" +
                 "    }\n" +
                 "}";
 
@@ -568,7 +568,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
         GlobalMavenConfig.get().setGlobalSettingsProvider(new FilePathGlobalSettingsProvider(mavenGlobalSettingsFile.getAbsolutePath()));
@@ -615,7 +615,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
 
@@ -666,7 +666,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
 
@@ -722,7 +722,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0', globalMavenSettingsFilePath: 'maven-global-settings.xml') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
 
@@ -766,7 +766,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0', mavenSettingsFilePath: 'maven-settings.xml') {\n" +
-                "        sh 'env && mvn help:effective-settings'\n" +
+                "        bat ' mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
 
@@ -810,7 +810,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
         GlobalMavenConfig.get().setSettingsProvider(new FilePathSettingsProvider(mavenSettingsFile.getAbsolutePath()));
@@ -857,7 +857,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
         GlobalConfigFiles.get().save(mavenSettingsConfig);
@@ -904,7 +904,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
         GlobalConfigFiles.get().save(mavenSettingsConfig);
@@ -960,7 +960,7 @@
                 "</project>'''\n" +
                 "\n" +
                 "    withMaven(maven: 'apache-maven-3.5.0', mavenSettingsConfig: 'maven-config-test-from-pipeline-attribute') {\n" +
-                "        sh 'mvn help:effective-settings'\n" +
+                "        bat 'mvn help:effective-settings'\n" +
                 "    }\n" +
                 "}\n";
 
@@ -987,7 +987,7 @@
                 "        node('master') {\n" +
                 "            git($/" + gitRepoRule.toString() + "/$)\n" +
                 "            withMaven() {\n" +
-                "                sh 'mvn package verify'\n" +
+                "                bat 'mvn package verify'\n" +
                 "            }\n" +
                 "        }\n" +
                 "    },\n" +
@@ -995,7 +995,7 @@
                 "        node('master') {\n" +
                 "            git($/" + gitRepoRule.toString() + "/$)\n" +
                 "            withMaven() {\n" +
-                "                sh 'mvn package verify'\n" +
+                "                bat 'mvn package verify'\n" +
                 "            }\n" +
                 "        }\n" +
                 "    })\n" +
diff --git a/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/publishers/InvokerRunsPublisherTest.java b/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/publishers/InvokerRunsPublisherTest.java
index 2f8c152..544e34e 100644
--- a/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/publishers/InvokerRunsPublisherTest.java
+++ b/jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/publishers/InvokerRunsPublisherTest.java
@@ -67,10 +67,10 @@
         Element projectsDirectoryElt = XmlUtils.getUniqueChildElementOrNull(pluginElt, "projectsDirectory");
 
         String reportsDirectory = invokerRunsPublisher.expandAndRelativize(reportsDirectoryElt, "reportsDirectory", invokerRunSucceedEvent, projectElt, workspace, listener);
-        assertThat(reportsDirectory, Matchers.is("target/invoker-reports"));
+        assertThat(reportsDirectory.replace('\\', '/'), Matchers.is("target/invoker-reports"));
         String projectsDirectory = invokerRunsPublisher.expandAndRelativize(projectsDirectoryElt, "projectsDirectory", invokerRunSucceedEvent, projectElt, workspace, listener);
-        assertThat(projectsDirectory, Matchers.is("src/it"));
+        assertThat(projectsDirectory.replace('\\', '/'), Matchers.is("src/it"));
         String cloneProjectsTo = invokerRunsPublisher.expandAndRelativize(cloneProjectsToElt, "cloneProjectsTo", invokerRunSucceedEvent, projectElt, workspace, listener);
-        assertThat(cloneProjectsTo, Matchers.is("target/it"));
+        assertThat(cloneProjectsTo.replace('\\', '/'), Matchers.is("target/it"));
     }
 }
```